lib/elastic_apm/sql/signature.rb

```diff
@@ -0,0 +1,139 @@
+# frozen_string_literal: true
+
+require 'elastic_apm/sql/tokenizer'
+
+module ElasticAPM
+  module Sql
+    # @api private
+    class Signature
+      include Tokens
+
+      def initialize(sql)
+        @sql = sql
+        @tokenizer = Tokenizer.new(sql)
+      end
+
+      def parse
+        @tokenizer.scan # until tokenizer.token != COMMENT
+
+        parsed = parse_tokens
+        return parsed if parsed
+
+        # If all else fails, just return the first token of the query.
+        parts = @sql.split
+        return '' unless parts.any?
+
+        parts.first.upcase
+      end
+
+      def self.parse(sql)
+        new(sql).parse
+      end
+
+      private
+
+      # rubocop:disable Metrics/CyclomaticComplexity
+      # rubocop:disable Metrics/PerceivedComplexity
+      def parse_tokens
+        t = @tokenizer
+
+        case t.token
+
+        when CALL
+          return unless scan_until IDENT
+          "CALL #{t.text}"
+
+        when DELETE
+          return unless scan_until FROM
+          return unless scan_token IDENT
+          table = scan_dotted_identifier
+          "DELETE FROM #{table}"
+
+        when INSERT, REPLACE
+          action = t.text
+          return unless scan_until INTO
+          return unless scan_token IDENT
+          table = scan_dotted_identifier
+          "#{action} INTO #{table}"
+
+        when SELECT
+          level = 0
+          while t.scan
+            case t.token
+            when LPAREN then level += 1
+            when RPAREN then level -= 1
+            when FROM
+              next unless level == 0
+              break unless scan_token IDENT
+              table = scan_dotted_identifier
+              return "SELECT FROM #{table}"
+            end
+          end
+
+        when UPDATE
+          # Scan for the table name. Some dialects allow option keywords before
+          # the table name.
+          return 'UPDATE' unless scan_token IDENT
+
+          table = t.text
+
+          period = false
+          first_period = false
+
+          while t.scan
+            case t.token
+            when IDENT
+              if period
+                table += t.text
+                period = false
+              end
+
+              unless first_period
+                table = t.text
+              end
+
+              # Two adjacent identifiers found after the first period. Ignore
+              # the secondary ones, in case they are unknown keywords.
+            when PERIOD
+              period = true
+              first_period = true
+              table += '.'
```

Comment:
For specifics like this, where we mostly know what to expect, ie. `/[[::alpha::]][[[::alpha:]]0-9_\.]*/`, I guess we could look into using more native StringScanner functionality.

Reply:
Similar to what I said above: I know we are trying to use a different strategy in this PR but perhaps a hybrid strategy makes the most sense for Ruby performance (?)

Reply:
```ruby
      # signature.rb
      def scan_dotted_identifier
        table = @tokenizer.text

        while @tokenizer.scan(/[[:alpha:]\.]+/)
          table += @tokenizer.text
        end

        return table
    end
```

```ruby
      # tokenizer.rb
      def scan(pattern = nil)
        if pattern
        @byte_start = scanner.pos
          result = scanner.scan_full(pattern, true, true)
          @byte_end = scanner.pos
          !!result
        else
          scanner.skip(SPACE)

        @byte_start = scanner.pos
          char = next_char

          return false unless char

          @token = next_token(char)

          true
        end
      end
```

Reply:
The above had no noticeable performance effects.

Reply:
I also think our benchmark data set doesn't have long enough statements to make a performance difference detectable...

Reply:
With the 24,000+ example set we get an improvement with the above:

Before:
```
============== Parsing 24594 examples ==============
              user     system      total        real
old:      0.319881   0.005052   0.324933 (  0.328290)
new:      2.060307   0.019711   2.080018 (  2.101040)
avg/ex:   0.000071   0.000001   0.000071 (  0.000072)
```

After:
```
============== Parsing 24594 examples ==============
              user     system      total        real
old:      0.328652   0.005300   0.333952 (  0.337635)
new:      1.920925   0.011025   1.931950 (  1.946499)
avg/ex:   0.000065   0.000000   0.000065 (  0.000065)
```

Yet we lose one example with a comment inside of the table names:
```sql
SELECT * FROM foo-- abc\n./*def*/bar
```
will now be `SELECT FROM foo` instead of `SELECT FROM foo.bar`. Possibly worth it as a sacrifice. Haven't come up with a way yet to have both.

Reply:
A few more runs bring the two closer together again. I'm not sure it's worth it right now.

Reply:
ok, thanks for investigating, it sounds like a change isn't worth it.